sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala

```diff
@@ -909,20 +909,66 @@ private case class GpuOrcFileFilterHandler(
      * the memory file will have a mapping of -1.
      *
      * @param fileIncluded indicator per column in the ORC file whether it should be included
-     * @return column mapping array
+     * @param fileSchema   ORC file schema
+     * @param readerSchema ORC schema for what will be read
+     * @return new column id mapping array and new column id -> old column id mapping array
      */
-    private def columnRemap(fileIncluded: Array[Boolean]): Array[Int] = {
-      var nextOutputColumnId = 0
-      val result = new Array[Int](fileIncluded.length)
-      fileIncluded.indices.foreach { i =>
-        if (fileIncluded(i)) {
-          result(i) = nextOutputColumnId
+    private def columnRemap(
+        fileIncluded: Array[Boolean],
+        fileSchema: TypeDescription,
+        readerSchema: TypeDescription): (Array[Int], Array[Int]) = {
+
+      // A column mapping for the new column id in the new re-constructing orc
+      val columnMapping = Array.fill[Int](fileIncluded.length)(-1)
+      // The first column is the top-level schema struct, always set it to 0
+      columnMapping(0) = 0
+      // The mapping for the new column id to the old column id which is used to get the encodings
+      val idMapping = Array.fill[Int](fileIncluded.length)(-1)
+      // The first column is the top-level schema struct, always set it to 0
+      idMapping(0) = 0
+
+      // the new column sequential id for the in-coming re-constructing orc
+      var nextOutputColumnId = 1
+
+      def setMapping(id: Int) = {
+        if (fileIncluded(id)) {
+          idMapping(nextOutputColumnId) = id
+          // change the column id for the new orc file
+          columnMapping(id) = nextOutputColumnId
           nextOutputColumnId += 1
-        } else {
-          result(i) = -1
         }
       }
-      result
+
+      //  Recursively update columnMapping and idMapping according to the TypeDescription id
+      // For now, we are only supporting struct and list nested type
+      def updateMapping(prefix: String, schema: TypeDescription, isRoot: Boolean = false): Unit = {
```

Comment:
It's been a while since I dug into ORC, so apologies if I'm missing something basic.  I'm not understanding why we need a recursive method.  IIUC all of the columns, whether they're nested or not, will appear in `fileIncluded`.  We're updating a flat-mapping of columns in `columnMapping` and `idMapping`, so that seems to be the case.

If we're just building a mapping back from the new column to the old column and keeping the existing mapping of old column ID to new column ID, why wouldn't we keep the existing function logic and simply update `idMapping` when we update `columnMapping`?  If we know enough to setup the forward-mapping for the column then we know how to set the reverse mapping as well.  Given it's flat-mapped on the input, I don't see why we need to special-case nesting here.

Reply:
the columnMapping should store the new columns' order following by **read schema** and it will be used when [setting the new footer](https://github.com/NVIDIA/spark-rapids/blob/branch-21.08/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala#L1031). But the original function logic always makes columnMapping store the old columns' order, which means we are pruning the columns according to the original orc file schema instead of read schema.

Given the below example,

Orc file schema: **struct<a:Int, b:string>**, the read scheam: **struct<b:string, a:int>**, then the fileIncluded=[true, true, true].

and the columnMapping in the original function will be **[0, 1, 2]**, which will follow the original ORC file schemal order, while the columnMapping with this PR will be **[0, 2, 1]**, which will follow the read schema order.

The recursive function is taking **read schema** as the parameter, and it will recursively lookup and set the mapping from the column id of read schema to the column id in the orc file schema.
 
I may not quite understand your question correctly, if that, sorry for that. @jlowe

Reply:
Thanks, @wbo4958.  I mistakenly read this as only creating a back-mapping with the new `idMapping` array and missed that it was also potentially reordering the existing `columnMapping`.